src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1302,6 +1302,7 @@ protected unsafe void WriteEventCore(int eventId, int eventDataCount, EventSourc
         [CLSCompliant(false)]
         protected unsafe void WriteEventWithRelatedActivityIdCore(int eventId, Guid* relatedActivityId, int eventDataCount, EventSource.EventData* data)
         {
+            WriteGuardEnter();
```

Comment:
The WriteGuardEnter happens outside the if statement and the WriteGuardLeave happens inside the if. I think there is a bug here if the following events happen

1. Thread A is writing an event
2. At the same time thread B calls Dispose, sets m_eventSourceEnabled to false and enters the spinloop waiting for thread A to finish
3. Now, after Dispose is called but before thread A finishes writing thread C also tries to write an event. It will increment the write, check if it is enabled and not execute the if statement, and therefore never decrement enabled
4. Now the spinloop will be guaranteed to time out because m_activeWritesCount will never be 0 again